### PR TITLE
testing strategy

### PR DESCRIPTION
**Pull request by jspaaks**
refs #132

Includes PR #149

### CODE REVIEW

File: setup.cfg [hunk 1]
```diff
@@ -24,6 +24,9 @@ branch = True
 source = howfairis
 
 [tool:pytest]
+python_files = test_*.py
+python_classes = Test*
+python_functions = test_*
```
Comment (sverhoeven) [inline setup.cfg:29]: Is this stilll needed, partly fixed in another PR (#147 ) with lower undersoring of modules

File: tests/contracts/config.py [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+    def test_default(self, mocked_config):
+        pass
+
+    @abstractmethod
+    def test_repo(self, mocked_config):
+        pass
+
+    @abstractmethod
+    def test_user(self, mocked_config):
+        pass
```
Comment (sverhoeven) [inline tests/contracts/config.py:16]: These properties have become private and should not be tested.
The other public props should be tested eg. force_checklist.

File: tests/gitlab/jspaaks/badge_test/__init__.py [hunk 3]
```diff
@@ -0,0 +1,6 @@
+from .mocker import mocker
+
+
+__all__ = [
+    "mocker"
+]
```
Comment (sverhoeven) [inline tests/gitlab/jspaaks/badge_test/__init__.py:6]: Can change the import in the testcases to `from .mocker import mocker` so we don't need this export.
```suggestion
```

File: tests/github/fair_software/badge/mocker.py [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+
+@pytest.fixture
+def mocker():
+    """This mock aims to reflect the state of the repository at
+    https://github.com/fair-software/badge/tree/b3f90ec9c2b1be604f482c2d9e46a9aeca3ee45a"""
+    with requests_mock.Mocker() as mocker:
+        mocker.get("http://github.com/fair-software/badge")
+        mocker.get("https://api.github.com/repos/fair-software/badge", json=dict(default_branch="master"))
+        mocker.get("https://raw.githubusercontent.com/fair-software/badge/master/.howfairis.yml", status_code=404)
+        return mocker
```
Comment (sverhoeven) [inline tests/github/fair_software/badge/mocker.py:13]: Use use same var name as func
```suggestion
    with requests_mock.Mocker() as m:
        m.get("http://github.com/fair-software/badge")
        m.get("https://api.github.com/repos/fair-software/badge", json=dict(default_branch="master"))
        mocker.get("https://raw.githubusercontent.com/fair-software/badge/master/.howfairis.yml", status_code=404)
        return m
```

Comment (sverhoeven) [inline tests/github/fair_software/badge/mocker.py]:
```suggestion
    with requests_mock.Mocker() as m:
        m.get("http://github.com/fair-software/badge")
        m.get("https://api.github.com/repos/fair-software/badge", json=dict(default_branch="master"))
        m.get("https://raw.githubusercontent.com/fair-software/badge/master/.howfairis.yml", status_code=404)
        return m
```

File: tests/gitlab/jspaaks/badge_test/mocker.py [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+
+@pytest.fixture
+def mocker():
+    """This mock aims to reflect the state of the repository at
+    https://gitlab.com/jspaaks/badge-test/-/tree/472713282fde87fa4c8e5116e46229c094c3c4ec"""
+    with requests_mock.Mocker() as mocker:
+        mocker.get("https://gitlab.com/jspaaks/badge-test")
+        mocker.get("https://gitlab.com/api/v4/projects/jspaaks%2Fbadge-test", json=dict(default_branch="master"))
+        mocker.get("https://gitlab.com/jspaaks/badge-test/-/raw/master/.howfairis.yml", status_code=404)
+        return mocker
```
Comment (sverhoeven) [inline tests/gitlab/jspaaks/badge_test/mocker.py:13]:
```suggestion
    with requests_mock.Mocker() as m:
        m.get("https://gitlab.com/jspaaks/badge-test")
        m.get("https://gitlab.com/api/v4/projects/jspaaks%2Fbadge-test", json=dict(default_branch="master"))
        m.get("https://gitlab.com/jspaaks/badge-test/-/raw/master/.howfairis.yml", status_code=404)
        return m
```

File: tests/github/fair_software/badge/test_config_no_args.py [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+
+    def test_default(self, mocked_config):
+        assert mocked_config.default == dict(force_repository=None, force_license=None, force_registry=None,
+                                             force_citation=None, force_checklist=None, include_comments=False)
+
+    def test_repo(self, mocked_config):
+        assert mocked_config.repo == dict()
+
+    def test_user(self, mocked_config):
+        assert mocked_config.user == dict()
```
Comment (sverhoeven) [inline tests/github/fair_software/badge/test_config_no_args.py:25]: Please replace with tests for the public interface of Config

**Review by sverhoeven — CHANGES_REQUESTED** [review]
Looks good, lets use the directory+mocker for each repo and use contract superclasses to enforce every test in every test case.

File: setup.cfg (hunk 1 above)
Comment (jspaaks) [reply setup.cfg:29]: removed

File: tests/contracts/config.py (hunk 2 above)
Comment (jspaaks) [reply tests/contracts/config.py:16]: removed these tests, updated contract, updated any tests that subsequently failed the contract

File: tests/gitlab/jspaaks/badge_test/__init__.py (hunk 3 above)
Comment (jspaaks) [reply tests/gitlab/jspaaks/badge_test/__init__.py:6]: done

File: tests/github/fair_software/badge/mocker.py (hunk 4 above)
Comment (jspaaks) [reply tests/github/fair_software/badge/mocker.py]: done

File: tests/github/fair_software/badge/test_config_no_args.py (hunk 6 above)
Comment (jspaaks) [reply tests/github/fair_software/badge/test_config_no_args.py:25]: done

**Review by sverhoeven — APPROVED** [review]
All suggestions have been applied
Thanks

**Comment by jspaaks** [comment]
yay thanks for reviewing

**Pull request merged**